### DISCUSSION TITLE
Modular?

### DISCUSSION THREAD

**Discussion by arbfay** [dragonflydb/dragonfly#74] [General]
Perf is nice, but can you extend the dragonfly?

A huge pain with redis is the excruciatingly difficult module system. RediSearch and RedisJson are quite good examples of what can be possible with modules, but C interfaces are awful and difficult to use, in 2022. 
So, hopefully, "modular" composes your definition of "well-designed" too?

**Comment by romange** [comment]
Do you know examples of modules that were implemented by someone who is not Redis Labs?

Even more specifically, modules/extensions are means to implement features that the maintainers of the project do not want or can not implement. Are there such features? What about LUA? does it provide the desired extensibility?

**Reply by arbfay** [reply]
Here's a list of modules: [https://redis.io/docs/modules/](https://redis.io/docs/modules/)
Quite nice to see things like TairHair or RediSQL.

It's not a feature request here, just trying to understand what you're planning with this. But I don't think the maintainers can realistically implement everything that could be built on top of their project.
PostgreSQL comes with a modular system too, which has given many interesting modules such as TimescaleDB.

And Lua scripts are no solution for things that modules cover (custom commands, data types, etc). For scripts/small programmes, I'd like to use wasm instead but that's another discussion.

**Comment by hpvd** [comment]
this topic is somehow related:
Use dragonfly with other parts of redis stack like RediSearch or RedisGraph
https://github.com/dragonflydb/dragonfly/discussions/1072

**Comment by kinthaiofficial** [comment]
The modularity question for a Redis-compatible server is interesting — there's real tension between compatibility (every Redis client assumes the same command set) and extensibility.

A few directions that could work:

**1. Module system via custom commands** (Redis Modules approach):
```
# Custom commands via COMMAND extension
MYMODULE.VECTORSEARCH key dim [...floats] k 10
MYMODULE.RATELIMIT key window_ms max_requests
```

Risk: breaks standard clients that don't know about custom commands.

**2. Lua scripting / Dragonfly Functions**:
```lua
-- Works with any Redis client via EVAL/FCALL
local key = KEYS[1]
local result = redis.call('ZRANGEBYSCORE', key, '-inf', ARGV[1])
-- custom post-processing logic here
return result
```

Lower risk, existing clients work fine, but limited to what Lua can express.

**3. Sidecar pattern** (what we use for agent memory caches):
Dragonfly handles Redis protocol, a sidecar handles domain-specific logic, talking to Dragonfly via standard Redis commands. Cleanest separation of concerns, no core changes needed.

For an agent memory use case (semantic caching, vector similarity over embeddings), we use the sidecar approach — the agent server does the embedding + similarity logic, uses Dragonfly for the actual data storage and expiry. Performance is fine because the embedding computation dominates anyway.

What's the specific module capability you're looking to add?

**Comment by romange** [comment]
Search  and JSON APIs were integrated into Dragonfly
